EventFilter/DTRawToDigi/plugins/DTuROSDigiToRaw.cc

```diff
@@ -0,0 +1,436 @@
+//-------------------------------------------------
+//
+//   Class: DTuROSDigiToRaw
+//
+//   L1 DT uROS Raw-to-Digi
+//
+//
+//
+//   Author :
+//   J. Troconiz  - UAM
+//   y el eterno retorno
+//
+//--------------------------------------------------
+
+#include "DTuROSDigiToRaw.h"
+
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "CondFormats/DTObjects/interface/DTReadOutMapping.h"
+#include "CondFormats/DataRecord/interface/DTReadOutMappingRcd.h"
+
+#include "DataFormats/MuonDetId/interface/DTWireId.h"
+#include "EventFilter/DTRawToDigi/interface/DTROChainCoding.h"
+
+#include <iostream>
+
+
+DTuROSDigiToRaw::DTuROSDigiToRaw(const edm::ParameterSet& pset) :  eventNum(0) {
+
+  produces<FEDRawDataCollection>();
+
+  DTDigiInputTag_ = pset.getParameter<edm::InputTag>("DTDigi_Source");
+
+  debug_ = pset.getUntrackedParameter<bool>("debug", false);
+
+  feds_ = pset.getUntrackedParameter<std::vector<int> >("feds", std::vector<int>());
+ 
+  nfeds_ = feds_.size();
+
+  Raw_token = consumes<DTDigiCollection>(DTDigiInputTag_);
+
+}
+
+
+DTuROSDigiToRaw::~DTuROSDigiToRaw(){}
+
+
+void DTuROSDigiToRaw::produce(edm::Event& e, const edm::EventSetup& c) {
+
+  FEDRawDataCollection data;
+
+  if (!fillRawData(e, c, data)) return;
+
+  auto fed_product = std::make_unique<FEDRawDataCollection>(data);
+
+  e.put(std::move(fed_product));
+
+}
+
+
+bool DTuROSDigiToRaw::fillRawData(edm::Event& e, const edm::EventSetup& c, FEDRawDataCollection& data) {
+
+  eventNum = e.id().event();
+
+  edm::Handle<DTDigiCollection> digis;
+  e.getByToken(Raw_token, digis);
+
+  edm::ESHandle<DTReadOutMapping> mapping;
+  c.get<DTReadOutMappingRcd>().get( mapping );
+
+  for (int w_i = 0; w_i < nfeds_; ++w_i) {
+    process(feds_[w_i], digis, mapping, data);
+  }
+  
+  return true;
+}
+
+
+void DTuROSDigiToRaw::process(int DTuROSFED,
+                              edm::Handle<DTDigiCollection> digis,
+                              edm::ESHandle<DTReadOutMapping> mapping,
+                              FEDRawDataCollection& data) {
+
+
+  clear();
+
+
+  //--> DTDigi analysis
+
+  DTDigiCollection::DigiRangeIterator dttax;
+  for (dttax = digis->begin(); dttax != digis->end(); ++dttax){
+    const DTDigiCollection::Range& dttar = (*dttax).second;
+    const DTLayerId dttal = (*dttax).first;
+    for (DTDigiCollection::const_iterator ta = dttar.first; ta != dttar.second; ++ta) {
+
+      int wheelId   = dttal.wheel();
+      int sectorId  = dttal.sector();
+      int stationId = dttal.station();
+      int slId      = dttal.superlayer();
+      int layerId   = dttal.layer();
+      int cellId    = (*ta).wire();
+      int tdcTime   = (*ta).countsTDC();
```

Comment:
This declaration should be better moved inside the if block

Reply:
done.